### PR TITLE
Remove Object#=~ and Object#!~

### LINKED ISSUE #5829: Object#=~ and Object#!~
The methods `Object#~=` and `Object#!~` are defined on `Object` but are not actually implemented (they return `nil` unless overwritten). This seems really odd, considering the intended purpose of matching a pattern is only useful to very select types. In the stdlib only `Regex` and `String` overwrite this method.
I don't think it makes sense to defined these methods for every `Object`, just to be able to use `=~` and `!~` as operators between arbitrary values (at least I guess this might be the reason for this).
Comparator methods for example are only defined on types including `Comparable` and that seems reasonable to me. You can only compare what's comparable. Equally, you should only be able to pattern match what's matchable. That is `String` and `Regex`.

### PR DESCRIPTION
**Pull request by straight-shoota**
Fixes #5829 

Overloads for `Regex#=~` are restricted to receive `String | Nil` and for `String#=~` to receive `Regex | Nil`. `#!~` is re-added as instance method in both classes accepting the same overload.

It could be up for debate to leave `Object#!~` to avoid having individual definitions in classes implementing `=~`. But this would end up in a similar issues as #3450

I'm also not entirely sure about the type safety of `Regex#=~` and `String#=~`. Should their argument really be nilable, or allow only `String`, `Regex` type, respectively? On the other hand, they could even receive any type and just return `nil` like it is the case right now.

### CODE REVIEW

**Comment by RX14** [comment]
I'm sure you'll realise why this was the way it was as you try to make CI pass.

File: src/spec/expectations.cr [hunk 1]
```diff
@@ -170,7 +170,7 @@ module Spec
     end
 
     def match(actual_value)
-      actual_value =~ @expected_value
+      actual_value.responds_to?(:=~) && actual_value =~ @expected_value
```
Comment (ysbaddaden) [inline src/spec/expectations.cr]: Not really a fix: you moved the "returns nil" logic from global operators to a specific use case.

The fact that the match expectation *is* called with non String nor RegExp objects may be proof that the global behavior is actually good?

**Comment by straight-shoota** [comment]
@ysbaddaden  True, I was just checking f there would be any other failures.

It seems it is just missing `Nil#=~(String | Regex)`. It would probably be fine to have this method for commutativity with `(String | Regex)#=~(Nil)`. I don't think it is necessary to have a default implementation for *any* type.

Although the issue is that `Nil#=~(other)` can't be restricted to `String | Regex` because other types might implement the matcher as well. If there is no type restriction at all, it defeats the purpose of stronger type safety. Possible solution would be a module `Matchable` (ugly) or a macro to check if `other` implements `#=~`.

Both are not optimal. I don't know if there is a better way. I still would like to avoid having `#=~` on every object.

**Comment by RX14** [comment]
sure nil is pretty common but it's not really specil compared to any other class. We should keep this as-is.

**Comment by asterite** [comment]
I also vote to close it. I'll explain (a bit more) why in #5829

**Pull request closed without merging**